cpp/src/arrow/python/datetime.cc

```diff
@@ -391,62 +391,165 @@ Result<std::string> PyTZInfo_utcoffset_hhmm(PyObject* pytzinfo) {
 Result<PyObject*> StringToTzinfo(const std::string& tz) {
   util::string_view sign_str, hour_str, minute_str;
   OwnedRef pytz;
-  RETURN_NOT_OK(internal::ImportModule("pytz", &pytz));
+  OwnedRef zoneinfo;
+  OwnedRef datetime;
 
+  if (internal::ImportModule("pytz", &pytz).ok()) {
+    RETURN_NOT_OK(internal::ImportModule("pytz", &pytz));
```

Comment:
```suggestion
  Status has_pytz = internal::ImportModule("pytz", &pytz)
  if (has_pytz.ok()) {
```

Something like this would avoid doing the ImportModule twice. Not that importing twice is really a problem, but this might be a bit cleaner.

(same for the other places where an ImportModule happens inside an `if`)

Reply:
Actually, just having it in the if statement as

```
if (internal::ImportModule("pytz", &pytz).ok()) {
    ...
```

is probably enough as well.